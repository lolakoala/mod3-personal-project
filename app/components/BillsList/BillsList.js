import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

class BillsList extends Component {
  getBillsToMap = () => {
    let billsToMap;
    const { usersHouse, currentUser, placeRendered } = this.props;
    const mybills = usersHouse.bills.filter(bill => {
      const usersOwe = bill.allUsersTotals.map(user => user.id);
      return usersOwe.includes(currentUser.id);
    });
    const billsDue = mybills.filter(bill => {
      return bill.allUsersTotals.find(user => user.id === currentUser.id).paid === false;
    });
    const billsNotPaidByAll = usersHouse.bills.filter(bill => {
      const usersNotPaid = bill.allUsersTotals.filter(user => user.paid === false);
      return usersNotPaid.length > 0;
    });
    if (placeRendered === '/') {
      billsToMap = billsDue;
    } else if (placeRendered === '/userlist') {
      billsToMap = mybills;
    } else if (placeRendered === '/houselist') {
      billsToMap = usersHouse.bills;
    } else if (placeRendered === 'summary') {
      billsToMap = billsNotPaidByAll;
    }
    return billsToMap;
  }

  getTitle = () => {
    const { placeRendered } = this.props;
    let title;
    if (placeRendered === '/') {
      title = 'Bills I Owe';
    } else if (placeRendered === '/userlist') {
      title = 'My Bills';
    } else if (placeRendered === '/houselist') {
      title = 'House Bills';
    } else if (placeRendered === 'summary') {
      title = 'Bills to Be Paid';
    }
    return title;
  }

  markBillPaid = (billId, userId, usersHouse) => {
    this.props.markBillPaid(billId, userId, usersHouse);
    this.forceUpdate();
  }

  getHouseBills = bills => {
    const { currentUser, usersHouse } = this.props;
    return bills.map(bill => {
      return (<div key={bill.parsedDuedate}>
        <Link to={`bills/${bill.id}`}>{bill.title}</Link>
        <p>{bill.duedate}</p>
        <p>{bill.total}</p>
        <div onClick={() => this.markBillPaid(bill.id, currentUser.id, usersHouse)}>
          {bill.allUsersTotals.find(user => user.id === currentUser.id).paid ? 'Paid' : 'Mark as Paid'}
        </div>
      </div>);
    });
  }

  getUserBills = bills => {
    const { currentUser, usersHouse } = this.props;
    return bills.map(bill => {
      const user = bill.allUsersTotals.find(user => user.id === currentUser.id);
      return (<div key={bill.parsedDuedate}>
        <Link to={`bills/${bill.id}`}>{bill.title}</Link>
        <p>{bill.duedate}</p>
        <p>{user.total}</p>
        <div onClick={() => this.markBillPaid(bill.id, currentUser.id, usersHouse)}>{user.paid ? 'Paid' : 'Mark as Paid'}</div>
      </div>);
    });
  }

  render() {
    const { placeRendered, usersHouse } = this.props;
    if (usersHouse.bills.length && usersHouse.bills[0].title !== 'fake') {
      const billsToMap = this.getBillsToMap();
      if (billsToMap.length) {
        return (
          <div>
            <h2>{this.getTitle()}</h2>
            <h4>Title</h4>
            <h4>Due Date</h4>
            <h4>{placeRendered === '/houselist' ? 'Total' :'My Total'}</h4>
            <h4>{placeRendered === '/houselist' ? 'All Paid' : 'Mark as Paid'}</h4>
            {placeRendered === '/houselist' ? this.getHouseBills(billsToMap) : null}
            {placeRendered === '/' ? this.getUserBills(billsToMap) : null}
            {placeRendered === '/userlist' ? this.getUserBills(billsToMap) : null}
            {placeRendered === 'summary' ? this.getHouseBills(billsToMap) : null}
          </div>
        );

      } else {
        return (
          <div></div>
        );
      }
    } else {
      return (
        <div></div>
      );
    }

  }
}

export default BillsList;

BillsList.propTypes = {
  usersHouse: PropTypes.object,
  currentUser: PropTypes.object,
  markBillPaid: PropTypes.func,
  placeRendered: PropTypes.string
};
