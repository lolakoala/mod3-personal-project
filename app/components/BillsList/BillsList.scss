.billslist h4,
.error {
  background-color: rgba(141, 132, 73, 0.7);
  color: #3d6aa2;
  font-family: 'Shadows Into Light', cursive;
  font-size: 20px;
  width: 250px;
  padding: 5px;
  text-align: center;
  height: 35px;
  border-radius: 5px;
  line-height: 35px;
  margin-top: 10px;
}

.billslist .bill {
  display: flex;
  height: 30px;
  width: 300px;
  border-radius: 5px;
  border: 2px solid rgba(141, 132, 73, 0.7);
  margin-top: 10px;
  line-height: 30px;
  font-family: 'BenchNine', sans-serif;
  position: relative;
}

.bill a {
  text-decoration: none;
  width: 75px;
  overflow: hidden;
  display: inline-block;
  margin: 2px 5px;
  background-color: #D83D59;
  border: 2px solid #D83D59;
  height: 25px;
  color: #F4CCE2;
  font-size: 20px;
  border-radius: 25px;
  text-align: center;
}

.bill a:hover {
  cursor: pointer;
  border: 2px dashed #D83D59;
  background-color: #F4CCE2;
  color: #D83D59;
}

.bill .due,
.bill .total {
  width: 60px;
  font-size: 15px;
  display: inline-block;
  margin: 2px 0px 2px 5px;
  text-align: center;
}

.bill .mark-paid {
  height: 100%;
  width: 75px;
  background-color: #eb6b3c;
  position: absolute;
  top: 0;
  right: 0;
  color: #F4CCE2;
  display: inline-block;
  font-size: 15px;
  padding-left: 10px;
}

.mark-paid:hover {
  cursor: pointer;
  background-color: #F4CCE2;
  color: #eb6b3c;
}
