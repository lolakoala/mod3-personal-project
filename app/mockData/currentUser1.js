const currentUser1 = {
  id : "RVde1m9BhkQBLZlTJNsFlQZXqP43",
  name: "Amanda Brenner",
  house: "Dog9"
};

export default currentUser1;
