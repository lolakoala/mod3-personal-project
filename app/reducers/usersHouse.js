import firebase from '../firebase.js';

const usersHouse = (state = {}, action) => {
  const housesRef = firebase.database().ref('houses');
  switch (action.type) {
  case 'CREATE_HOUSE':
    housesRef.push(action.house);
    return action.house;
  case 'GET_HOUSE':
    const updatedHouse = Object.assign(action.usersHouse, { users: [...action.usersHouse.users, action.user] });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(updatedHouse);
    return updatedHouse;
  case 'LOGIN_SUCCESS':
    return action.usersHouse;
  case 'SIGNOUT':
    return {};
  case 'LEAVE_HOUSE':
    const deleteUser = action.usersHouse.users.filter(user => user.id !== action.currentUser.id);
    const houseWithoutUser = Object.assign(action.usersHouse, { users: deleteUser });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(houseWithoutUser);
    return {};
  case 'ADD_BILL':
    const takeOutFakeBills = action.usersHouse.bills.filter(bill => bill.title !== 'fake');
    const houseWithBill = Object.assign(action.usersHouse, { bills: [...takeOutFakeBills, action.bill] });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(houseWithBill);
    return houseWithBill;
  case 'ADD_BULLETIN':
    const takeOutFakeBulletins = action.usersHouse.bulletins.filter(bulletin => bulletin.title !== 'fake');
    const houseWithBulletin = Object.assign(action.usersHouse, { bulletins: [...takeOutFakeBulletins, action.bulletin] });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(houseWithBulletin);
    return houseWithBulletin;
  case 'ADD_CHORE':
    const takeOutFakeChore = action.usersHouse.chores.filter(chore => chore.title !== 'fake');
    const houseWithChore = Object.assign(action.usersHouse, { chores: [...takeOutFakeChore, action.chore] });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(houseWithChore);
    return houseWithChore;
  case 'ADD_READER':
    const oldBulletin = action.usersHouse.bulletins.find(bulletin => bulletin.id === action.bulletinId);
    if (oldBulletin.hasRead.includes(action.userId)) {
      return action.usersHouse;
    }
    const newBulletin = Object.assign(oldBulletin, { hasRead: [...oldBulletin.hasRead, action.userId] });
    const updateBulletins = action.usersHouse.bulletins.filter(bulletin => {
      return bulletin.id !== action.bulletinId;
    });
    const updateWithBulletin = Object.assign(action.usersHouse, { bulletins: [...updateBulletins, newBulletin] });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(updateWithBulletin);
    return updateWithBulletin;
  case 'MARK_PAID':
    const oldBill = action.usersHouse.bills.find(bill => bill.id === action.billId);
    const user = oldBill.allUsersTotals.find(user => user.id === action.userId);
    if (user.paid === true) {
      return action.usersHouse;
    }
    const newUser = Object.assign(user, { paid: true });
    const newAllUsersTotals = oldBill.allUsersTotals.filter(user => user.id !== action.userId);
    const newBill = Object.assign(oldBill, { allUsersTotals: [...newAllUsersTotals, newUser] });
    const updatedBills = action.usersHouse.bills.filter(bill => {
      return bill.id !== action.billId;
    });
    const updateWithBill = Object.assign(action.usersHouse, { bills: [...updatedBills, newBill] });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(updateWithBill);
    return updateWithBill;
  default:
    return state;
  }
};

export default usersHouse;
