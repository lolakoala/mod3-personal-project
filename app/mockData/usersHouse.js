import currentUser1 from './currentUser1.js';
import currentUser2 from './currentUser2.js';

const bill1 = {
  allUsersTotals: [
    {
      id: currentUser1.id,
      paid: false,
      total: 50
    },
    {
      id: currentUser2.id,
      paid: false,
      total: 50
    }
  ],
  datePosted: "11/01/2017",
  details: "We used a lot of water.",
  duedate: "11/9/17",
  id: "b3019762705779",
  parsedDuedate: 1510210800000,
  postedBy: currentUser1,
  title: "Water Bill",
  total: "100"
};

const bulletin1 = {
  datePosted: "11/01/2017",
  details: "The dogs got out through a hole in the fence yesterday.",
  hasRead: [currentUser1.id],
  id: 1509567098747,
  postedBy: currentUser1,
  title: "Don't leave dogs out unattended."
};

const chore1 = {
  assignedTo: currentUser1.id,
  datePosted: "10/31/2017",
  details: "Empty the dishwasher.",
  done: false,
  id: "c1509485241076",
  postedBy: currentUser1,
  title: "Dishes",
  urgency: "low"
};

const usersHouse = {
  houseKey: "-KxoIBWy5UYRscFYdWty",
  bills: [bill1],
  bulletins: [bulletin1],
  chores: [chore1],
  houseCode: "woof9",
  houseName: "Dog9",
  users: [currentUser1, currentUser2]
};

export default usersHouse;
