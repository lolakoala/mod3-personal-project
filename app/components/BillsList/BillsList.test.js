import React from "react";
import BillsList from './BillsList.js';
import { shallow } from "enzyme";
import usersHouse from '../../mockData/usersHouse.js';
import currentUser1 from '../../mockData/currentUser1.js';


describe('BillsList component', () => {
  const mockFn = jest.fn();
  const wrapper = shallow(<BillsList usersHouse={usersHouse}
    currentUser={currentUser1}
    markBillPaid={mockFn}
    placeRendered={'/houselist'}
    searchValue={''}/>);

  it('should exist', () => {
    expect(wrapper).toBeDefined();
  });

  it('should have a title', () => {
    const title = wrapper.find('.billslist-title');

    expect(title.text()).toEqual('House Bills');
  });

  it('should have a title column header', () => {
    const titleHeader = wrapper.find('.title-header');

    expect(titleHeader.text()).toEqual('Title');
  });

  it('should have a duedate column header', () => {
    const duedateHeader = wrapper.find('.duedate-header');

    expect(duedateHeader.text()).toEqual('Due Date');
  });

  it('should have a total header', () => {
    const totalHeader = wrapper.find('.total-header');

    expect(totalHeader.text()).toEqual('Total');
  });

  it('should have a paid header', () => {
    const paidHeader = wrapper.find('.paid-header');

    expect(paidHeader.text()).toEqual('All Paid');
  });

  it('should have display all bills of the house', () => {
    const bills = wrapper.find('.bill');

    expect(bills.length).toEqual(usersHouse.bills.length);
  });

  it('should fire action', () => {
    const markPaid = wrapper.find('.mark-paid').first();

    markPaid.simulate('click');
    expect(mockFn).toHaveBeenCalledTimes(1);
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
