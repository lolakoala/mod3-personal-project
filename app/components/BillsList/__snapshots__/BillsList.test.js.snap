// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BillsList component should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <BillsList
    currentUser={
      Object {
        "house": "Dog9",
        "id": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
        "name": "Amanda Brenner",
      }
    }
    markBillPaid={[Function]}
    placeRendered="/houselist"
    searchValue=""
    usersHouse={
      Object {
        "bills": Array [
          Object {
            "allUsersTotals": Array [
              Object {
                "id": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
                "paid": false,
                "total": 50,
              },
              Object {
                "id": "cFNdorl5ceRyMr9XGvPFRVEcw7I3",
                "paid": false,
                "total": 50,
              },
            ],
            "datePosted": "11/01/2017",
            "details": "We used a lot of water.",
            "duedate": "11/9/17",
            "id": "b3019762705779",
            "parsedDuedate": 1510210800000,
            "postedBy": Object {
              "house": "Dog9",
              "id": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
              "name": "Amanda Brenner",
            },
            "title": "Water Bill",
            "total": "100",
          },
        ],
        "bulletins": Array [
          Object {
            "datePosted": "11/01/2017",
            "details": "The dogs got out through a hole in the fence yesterday.",
            "hasRead": Array [
              "RVde1m9BhkQBLZlTJNsFlQZXqP43",
            ],
            "id": 1509567098747,
            "postedBy": Object {
              "house": "Dog9",
              "id": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
              "name": "Amanda Brenner",
            },
            "title": "Don't leave dogs out unattended.",
          },
        ],
        "chores": Array [
          Object {
            "assignedTo": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
            "datePosted": "10/31/2017",
            "details": "Empty the dishwasher.",
            "done": false,
            "id": "c1509485241076",
            "postedBy": Object {
              "house": "Dog9",
              "id": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
              "name": "Amanda Brenner",
            },
            "title": "Dishes",
            "urgency": "low",
          },
        ],
        "houseCode": "woof9",
        "houseKey": "-KxoIBWy5UYRscFYdWty",
        "houseName": "Dog9",
        "users": Array [
          Object {
            "house": "Dog9",
            "id": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
            "name": "Amanda Brenner",
          },
          Object {
            "house": "Dog9",
            "id": "cFNdorl5ceRyMr9XGvPFRVEcw7I3",
            "name": "Lola Koala",
          },
        ],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        null,
        null,
        Array [
          <div
            className="bill"
          >
            <Link
              replace={false}
              to="bills/b3019762705779"
            >
              Water Bill
            </Link>
            <p
              className="due"
            >
              Due: 11/9/17
            </p>
            <p
              className="total"
            >
              Total: 100
            </p>
            <p
              className="mark-paid"
              onClick={[Function]}
            >
              Mark Paid
            </p>
          </div>,
        ],
        null,
        null,
        null,
        null,
      ],
      "className": "billslist",
    },
    "ref": null,
    "rendered": Array [
      null,
      null,
      Object {
        "instance": null,
        "key": "1510210800000",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Link
              replace={false}
              to="bills/b3019762705779"
            >
              Water Bill
            </Link>,
            <p
              className="due"
            >
              Due: 11/9/17
            </p>,
            <p
              className="total"
            >
              Total: 100
            </p>,
            <p
              className="mark-paid"
              onClick={[Function]}
            >
              Mark Paid
            </p>,
          ],
          "className": "bill",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": null,
            "nodeType": "class",
            "props": Object {
              "children": "Water Bill",
              "replace": false,
              "to": "bills/b3019762705779",
            },
            "ref": null,
            "rendered": "Water Bill",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "children": "Due: 11/9/17",
              "className": "due",
            },
            "ref": null,
            "rendered": "Due: 11/9/17",
            "type": "p",
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "children": "Total: 100",
              "className": "total",
            },
            "ref": null,
            "rendered": "Total: 100",
            "type": "p",
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "children": "Mark Paid",
              "className": "mark-paid",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Mark Paid",
            "type": "p",
          },
        ],
        "type": "div",
      },
      null,
      null,
      null,
      null,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          null,
          null,
          Array [
            <div
              className="bill"
            >
              <Link
                replace={false}
                to="bills/b3019762705779"
              >
                Water Bill
              </Link>
              <p
                className="due"
              >
                Due: 11/9/17
              </p>
              <p
                className="total"
              >
                Total: 100
              </p>
              <p
                className="mark-paid"
                onClick={[Function]}
              >
                Mark Paid
              </p>
            </div>,
          ],
          null,
          null,
          null,
          null,
        ],
        "className": "billslist",
      },
      "ref": null,
      "rendered": Array [
        null,
        null,
        Object {
          "instance": null,
          "key": "1510210800000",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Link
                replace={false}
                to="bills/b3019762705779"
              >
                Water Bill
              </Link>,
              <p
                className="due"
              >
                Due: 11/9/17
              </p>,
              <p
                className="total"
              >
                Total: 100
              </p>,
              <p
                className="mark-paid"
                onClick={[Function]}
              >
                Mark Paid
              </p>,
            ],
            "className": "bill",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": null,
              "nodeType": "class",
              "props": Object {
                "children": "Water Bill",
                "replace": false,
                "to": "bills/b3019762705779",
              },
              "ref": null,
              "rendered": "Water Bill",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "children": "Due: 11/9/17",
                "className": "due",
              },
              "ref": null,
              "rendered": "Due: 11/9/17",
              "type": "p",
            },
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "children": "Total: 100",
                "className": "total",
              },
              "ref": null,
              "rendered": "Total: 100",
              "type": "p",
            },
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "children": "Mark Paid",
                "className": "mark-paid",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Mark Paid",
              "type": "p",
            },
          ],
          "type": "div",
        },
        null,
        null,
        null,
        null,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
